roadmap/implementers-guide/src/node/approval/approval-voting.md

```diff
@@ -0,0 +1,234 @@
+    tranches: Vec<TrancheEntry>, // sorted ascending by tranche number.
+    backing_group: GroupIndex,
+    // When the next wakeup for this entry should occur. This is either to
+    // check a no-show or to check if we need to broadcast an assignment.
+    next_wakeup: Tick,
+    our_assignment: Option<OurAssignment>,
+    assignments: Bitfield, // n_validators bits
+    approved: bool,
+}
+
+struct CandidateEntry {
+    candidate: CandidateReceipt,
+    session: SessionIndex,
+    // Assignments are based on blocks, so we need to track assignments separately
+    // based on the block we are looking at.
+    block_assignments: HashMap<Hash, ApprovalEntry>,
+    approvals: Bitfield, // n_validators bits
+}
+
+struct BlockEntry {
+    block_hash: Hash,
+    session: SessionIndex,
+    slot: SlotNumber,
+    received_late_by: Duration,
+    // random bytes derived from the VRF submitted within the block by the block
+    // author as a credential and used as input to approval assignment criteria.
+    relay_vrf_story: [u8; 32],
+    // The candidates included as-of this block and the index of the core they are
+    // leaving. Sorted ascending by core index.
+    candidates: Vec<(CoreIndex, Hash)>,
+    // A bitfield where the i'th bit corresponds to the i'th candidate in `candidates`.
+    // The i'th bit is `true` iff the candidate has been approved in the context of
+    // this block. The block can be considered approved if the bitfield is equal to !0.
+    approved_bitfield: Bitfield,
+    rotation_offset: GroupIndex,
+    children: Vec<Hash>,
+}
+
+// slot_duration * 2 + DelayTranche gives the number of delay tranches since the
+// unix epoch.
+type Tick = u64;
+
+struct TrackerEntry 
+
+struct StoredBlockRange(BlockNumber, BlockNumber)
+```
+
+In the schema, we map
+
+```
+"StoredBlocks" => StoredBlockRange
+BlockNumber => Vec<BlockHash>
+BlockHash => BlockEntry
+CandidateHash => CandidateEntry
+```
+
+## Logic
+
+In-memory state:
+
+```rust
+struct ApprovalVoteRequest {
+  validator_index: ValidatorIndex,
+  block_hash: Hash,
+  candidate_index: u32,
+}
+
+struct State {
+    earliest_session: SessionIndex,
+    session_info: Vec<SessionInfo>,
+    keystore: KeyStorePtr,
+    wakeups: BTreeMap<Tick, Vec<(Hash, Hash)>>, // Tick -> [(Relay Block, Candidate Hash)]
+    
+    // These are connected to each other.
+    approval_vote_tx: mpsc::Sender<ApprovalVoteRequest>,
+    approval_vote_rx: mpsc::Receiver<ApprovalVoteRequest>,
+}
+```
+
+[`SessionInfo`](../../runtime/session_info.md)
+
+On start-up, we clear everything currently stored by the database. This is done by loading the `StoredBlockRange`, iterating through each block number, iterating through each block hash, and iterating through each candidate referenced by each block. Although this is O(n^3), we don't expect to have more than a few unfinalized blocks at any time and in extreme cases, a few thousand. The clearing operation should be relatively fast as a result.
```

Comment:
It's not really `O(n^3)`, because the number of stored blocks in the range, the number of hashes for each block, and the number of candidates referenced per block all have differing cardinality. AFAICT these are entirely distinct, which makes the bound `O(n*o*p)`, and I (possibly naively) expect `o` and `p` to remain pretty small all the time.